skfp/similarities/tanimoto.py

```diff
@@ -0,0 +1,229 @@
+            f"Both A and B must be of the same type: either numpy.ndarray "
+            f"or scipy.sparse.csr_matrix, got {type(A)} and {type(B)}"
+        )
+
+
+def binary_tanimoto_distance(
+    A: Union[np.ndarray, csr_matrix], B: Union[np.ndarray, csr_matrix]
+) -> float:
+    """
+    Computes the Tanimoto distance for binary data between two input arrays or sparse matrices.
+
+    Parameters
+    ----------
+    A : np.darray or scipy.sparse.csr_matrix
+        First binary input array or sparse matrix.
+    B : np.darray or scipy.sparse.csr_matrix
+        Second binary input array or sparse matrix.
+
+    Returns
+    ----------
+    float
+        Tanimoto distance between A and B.
+    """
+
+    return 1 - binary_tanimoto_similarity(A, B)
+
+
+def count_tanimoto_similarity(
+    A: Union[np.ndarray, csr_matrix], B: Union[np.ndarray, csr_matrix]
+) -> float:
+    """
+    Computes the Tanimoto similarity [1]_ for continuous data between two input arrays
+    or sprase matrices. Calls suitable private method based on input type.
+
+    Parameters
+    ----------
+    A : np.darray or scipy.sparse.csr_matrix
+        First binary input array or sparse matrix.
+    B : np.darray or scipy.sparse.csr_matrix
+        Second binary input array or sparse matrix.
+
+    Returns
+    ----------
+    tanimoto_similarity : float
+        Tanimoto similarity between A and B.
+
+    References
+    ----------
+    .. [1] `Why is Tanimoto index an appropriate choice for fingerprint-based similarity calculations?
+    https://jcheminf.biomedcentral.com/articles/10.1186/s13321-015-0069-3`_
+    """
+    if isinstance(A, csr_matrix) and isinstance(B, csr_matrix):
+        return _count_scipy_tanimoto(A, B)
+    elif isinstance(A, np.ndarray) and isinstance(B, np.ndarray):
+        return _count_numpy_tanimoto(A, B)
+    else:
+        raise TypeError(
+            f"Both A and B must be of the same type: either numpy.ndarray "
+            f"or scipy.sparse.csr_matrix, got {type(A)} and {type(B)}"
+        )
+
+
+def count_tanimoto_distance(
+    A: Union[np.ndarray, csr_matrix], B: Union[np.ndarray, csr_matrix]
+) -> float:
+    """
+    Computes the Tanimoto distance for binary data between two input arrays or sparse matrices.
+
+    Parameters
+    ----------
+    A : np.darray or scipy.sparse.csr_matrix
+        First binary input array or sparse matrix.
+    B : np.darray or scipy.sparse.csr_matrix
+        Second binary input array or sparse matrix.
+
+    Returns
+    ----------
+    float
+        Tanimoto distance between A and B.
+    """
+    return 1 - count_tanimoto_similarity(A, B)
+
+
+def _binary_numpy_tanimoto(A: np.ndarray, B: np.ndarray) -> float:
+    """
+    Calculate the Tanimoto similarity between two binary numpy arrays.
+    """
+    _check_nan(A)
+    _check_nan(B)
+
+    if np.sum(A) == 0 and np.sum(B) == 0:
+        return 1.0
+
+    dot_product_AB: float = np.dot(A, B)
+    dot_product_AA: float = np.sum(A)
```

Comment:
Not sure, but as far as I remember, dot product of the same vectors is sum of squares. Change it or rename variable please